Mathlib/LinearAlgebra/Matrix/SVD/svdReindex.lean

```diff
@@ -0,0 +1,100 @@
+/-
+Copyright (c) 2023 Mohanad ahmed. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mohanad Ahmed
+-/
+
+import Mathlib.LinearAlgebra.Matrix.PosDef
+import Mathlib.Data.Matrix.Rank
+
+/-! # Reindexing using Non-zero/Zero Partition Eigenvalues of AAᴴ and AᴴA
+
+Given a Fin M × Fin N we wish to define equivlances that allow us to move between the following
+representaions smoothly. Let R be the rank
+  Fin N ≃ (Fin R ⊕ Fin (N - R)) which can be achievd through
+  Fin N ≃ {Non-zero Eigs} ⊕ (Zero Eigs) ≃ Fin R ⊕ Fin (N - R)
+
+  Fin M ≃ (Fin R ⊕ Fin (M - R)) which can be achievd through
+  Fin M ≃ {Non-zero Eigs} ⊕ (Zero Eigs) ≃ Fin R ⊕ Fin (M - R)
+
+  Note that we know R ≤ MIN(M, N) -/
+variable {𝕂: Type}[IsROrC 𝕂][DecidableEq 𝕂]
+variable {M N: ℕ}
+
+open Matrix BigOperators
+
+namespace Matrix
+
+open scoped ComplexOrder
+
+lemma rank_eq_card_pos_eigs_conj_transpose_mul_self (A: Matrix (Fin M) (Fin N) 𝕂) :
+    A.rank = Fintype.card {i // (isHermitian_transpose_mul_self A).eigenvalues i ≠ 0} := by
+  rw[← rank_conjTranspose_mul_self, IsHermitian.rank_eq_card_non_zero_eigs]
+
+lemma rank_eq_card_pos_eigs_self_mul_conj_transpose (A: Matrix (Fin M) (Fin N) 𝕂) :
+    A.rank = Fintype.card {i // (isHermitian_mul_conjTranspose_self A).eigenvalues i ≠ 0} := by
+  rw[← rank_self_mul_conjTranspose, IsHermitian.rank_eq_card_non_zero_eigs]
+
+noncomputable def er (A: Matrix (Fin M) (Fin N) 𝕂) :
```

Comment:
I don't understand the name of this def'n nor the names of the def'ns that follow it. Could you please expand the names?
Also, please add docstrings to all definitions.

Reply:
OK added some docstrings to the definitions